code/Model/RedirectorPageController.php

```diff
@@ -11,19 +12,38 @@ class RedirectorPageController extends PageController
 {
     private static $allowed_actions = ['index'];
 
+    /**
+     * Should respond with HTTP 404 if the page or file being redirected to is missing
+     */
+    private static bool $missing_redirect_is_404 = true;
+
     /**
      * Check we don't already have a redirect code set
      *
      * @param  HTTPRequest $request
      * @return \SilverStripe\Control\HTTPResponse
+     * @throws HTTPResponse_Exception
      */
     public function index(HTTPRequest $request)
     {
         /** @var RedirectorPage $page */
         $page = $this->data();
+
+        // Redirect if we can
         if (!$this->getResponse()->isFinished() && $link = $page->redirectionLink()) {
-            $this->redirect($link, 301);
+            return $this->redirect($link, 301);
         }
+
+        // Respond with 404 if redirecting to a missing file or page
+        if (($this->RedirectionType === 'Internal' && !$page->LinkTo()?->exists())
+            || ($this->RedirectionType === 'File' && !$page->LinkToFile()?->exists())
+        ) {
+            if (static::config()->get('missing_redirect_is_404')) {
+                return $this->httpError(404);
```

Comment:
```suggestion
                $this->httpError(404);
```
RequestHandler::httpError() throws an exception, doesn't return anything

Reply:
I don't think this really matters either way - but since I need to touch this PR anyway to add an extra test I'll also change this.
I personally like the `return` because it makes it super clear to the reader that this line will stop executing the rest of the method, but it's not really important.

Reply:
Done